File: samples/dotnet-aspnet/src/XmlIngestService.cs
using System.Xml;

namespace Reporting;

/// <summary>
/// Ingestion of externally supplied report definitions.
///
/// Intentionally vulnerable, see cases/xml-resolver-helper.
/// </summary>
public class XmlIngestService
{
    /// <summary>
    /// Build the document used for ingestion.
    ///
    /// Modern .NET leaves XmlResolver null, which is safe. Assigning an XmlUrlResolver turns that off
    /// and restores external entity resolution, so the hardening has been deliberately undone here.
    /// </summary>
    private XmlDocument NewDocument()
    {
        var document = new XmlDocument();

        // VULN: re-enables external entity resolution that the platform disables by default.
        document.XmlResolver = new XmlUrlResolver();

        return document;
    }

    public string Ingest(string xml)
    {
        XmlDocument document = NewDocument();

        document.LoadXml(xml);

        return document.DocumentElement?.Name ?? string.Empty;
    }
}

File: samples/java-spring/src/main/java/com/example/reporting/XmlIngestService.java
package com.example.reporting;

import java.io.ByteArrayInputStream;
import java.nio.charset.StandardCharsets;
import javax.xml.parsers.DocumentBuilder;
import javax.xml.parsers.DocumentBuilderFactory;
import org.springframework.stereotype.Service;
import org.w3c.dom.Document;

/**
 * Ingestion of externally supplied report definitions.
 *
 * Intentionally vulnerable, see cases/xxe-parser-helper.
 */
@Service
public class XmlIngestService {

    /**
     * Build the parser used for ingestion.
     *
     * The factory is left at its defaults, which permit DTD processing and external entity resolution.
     * The dangerous configuration is here, while the untrusted document is supplied by the caller.
     */
    private DocumentBuilder newBuilder() throws Exception {
        // VULN: no feature is disabled, so a DOCTYPE in the payload can read local files or reach internal hosts.
        DocumentBuilderFactory factory = DocumentBuilderFactory.newInstance();

        return factory.newDocumentBuilder();
    }

    public String ingest(String xml) throws Exception {
        DocumentBuilder builder = newBuilder();

        Document document = builder.parse(new ByteArrayInputStream(xml.getBytes(StandardCharsets.UTF_8)));

        return document.getDocumentElement().getNodeName();
    }
}
